Separate CGB Infrared section; clarify IR Receiver Adaptation (#327)
gbdev/pandocs: src/CGB_Infrared.md

# CGB Infrared

The GBC comes with an infrared port on the very top of the handheld,
which can be used to send and receive IR signals via its two separate
diodes. This functionality is not limited to other GBCs - it can be

> used to communicate with external devices as well.

Reviewer: ```suggestion
used to communicate with other kinds of devices as well.
```
Maybe?

Note that the GBA does not include an infrared port and, as such, this
method of communication is unavailable in its CGB compatibility mode.

## FF56 - RP - Infrared Communications Port

This register allows data to be transmitted and received through the
infrared port.

> Bit | Name             | Usage notes                     | R/W?

Reviewer: The format of these is being discussed in #318, feel free to weigh in there if this is the style you prefer (maybe not, since it looks like you may have sliced this section from the original docs verbatim).

In the meantime, I'd like this be kept consistent with the style used throughout the rest of the document ([such as here](https://gbdev.io/pandocs/STAT.html#ff41---stat-lcd-status-rw)).

Author: I've based it on the LCDC page, as I assumed this is the current "desired" style. The original docs's format differs.

Reviewer: Yeah, we know the style is not used consistently, hence that issue. :/

Reviewer: (By the way, I'm picking this one specifically as I believe it's the best compromise until the new syntax is implemented.)

----|------------------|---------------------------------|-------------
6-7 | Data read enable | 0=Disable, 3=Enable             | Read/Write
 1  | Read data        | 0=Receiving IR Signal, 1=Normal | Read Only
 0  | Write data       | 0=LED Off, 1=Led On             | Read/Write

In order to read data, both Bit 6 and 7 must be set. Note that if Bit 0

> is not cleared, the CGB will end up reading its own IR signal.

Reviewer: ```suggestion
is also set, the CGB will end up reading its own IR signal.
```

Reviewer 2: The wording here ("not cleared") implies that, for proper usage, you _should_ clear the bit. This is correct, so maybe the stronger wording is helpful.

Reviewer: It's possible to write once to enable reading, and once again to disable the emitter, for example. So maybe "while bit 0 is also set, ..."?

> Both sending and receiving IR data increase power consumption - when not

Reviewer 3: ```suggestion
Since sending and receiving IR data increase power consumption - when not
```

> used, RP should be reset to 00h.

Reviewer: ```suggestion
used, RP should be reset to $00.
```

## IR Receiver Adaptation

The IR sensor in the GBC adapts to the current level of infrared light
as a way to handle potential light pollution. In other words, if the
GBC receives a sustained IR signal beyond a certain amount of time,
eventually the sensor treats this as a new "normal" level of light,
which causes Bit 1 of RP to go back to 1. This is called the signal
"fade" because it may appear as if the signal disappears.

The following should be kept in mind while working with the infrared
port:

* IR communication protocols should be designed with this effect in mind.

>   For example, a TV remote control may send a series of 32 LED ON/OFF

Reviewer: ```suggestion
  For example, some TV remotes send a series of 32 LED ON/OFF
```
(Note sure if that's *actually* true, but I think it doesn't matter much while making the wording less confusing.)

  pulses (length 10us ON, 17.5us OFF each) instead of a permanent 880us
  LED ON signal.
* When enabling data read in the RP register under brighter light
  conditions, one may initially read 0 from Bit 1, implying that an IR
  signal is being received. As such, a short amount of time should pass

>   between writing 0xC0 to RP and attempting to read an IR signal. (TODO:

Reviewer 3: ```suggestion
  between writing $C0 to RP and attempting to read an IR signal. (TODO:
```

See https://github.com/gbdev/pandocs/wiki/Document-Style
